opm/input/eclipse/Schedule/Well/WCYCLE.cpp

```diff
@@ -0,0 +1,205 @@
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <opm/input/eclipse/Schedule/Well/WCYCLE.hpp>
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+
+#include <opm/input/eclipse/Deck/DeckRecord.hpp>
+#include <opm/input/eclipse/Parser/ParserKeywords/W.hpp>
+#include <opm/input/eclipse/Schedule/Well/WellMatcher.hpp>
+#include <opm/input/eclipse/Units/Units.hpp>
+
+#include <fmt/format.h>
+
+namespace Opm {
+
+WCYCLE WCYCLE::serializationTestObject()
+{
+    WCYCLE result;
+
+    result.entries_.emplace("W1", Entry{1.0, 2.0, 3.0, 4.0, false});
+    result.entries_.emplace("W2", Entry{5.0, 6.0, 7.0, 8.0, true});
+
+    return result;
+}
+
+bool WCYCLE::Entry::operator==(const Entry& that) const
+{
+    return this->on_time == that.on_time
+        && this->off_time == that.off_time
+        && this->startup_time == that.startup_time
+        && this->max_time_step == that.max_time_step
+        && this->controlled_time_step == that.controlled_time_step;
+}
+
+void WCYCLE::addRecord(const DeckRecord& record)
+{
+    const std::string name = record.getItem<ParserKeywords::WCYCLE::WELL>().getTrimmedString(0);
+    entries_.emplace(name, Entry {
+        record.getItem<ParserKeywords::WCYCLE::ON_TIME>().getSIDouble(0),
+        record.getItem<ParserKeywords::WCYCLE::OFF_TIME>().getSIDouble(0),
+        record.getItem<ParserKeywords::WCYCLE::START_TIME>().getSIDouble(0),
+        record.getItem<ParserKeywords::WCYCLE::MAX_TIMESTEP>().getSIDouble(0),
+        record.getItem<ParserKeywords::WCYCLE::CONTROLLED_TIMESTEP>().getTrimmedString(0) == "YES"
+    });
+}
+
+bool WCYCLE::operator==(const WCYCLE& that) const
+{
+    return this->entries_ == that.entries_;
+}
+
+double WCYCLE::nextTimeStep(const double current_time,
+                            const double dt,
+                            const WellMatcher& wmatch,
+                            const WellTimeMap& open_times,
+                            const WellTimeMap& close_times,
+                            std::function<bool(const std::string&)> opens_this_step) const
+{
+    double next_dt = dt;
+    for (const auto& [name, wce]: entries_) {
+        if (wce.off_time > 0 && wce.controlled_time_step) {
+            for (const auto& w : wmatch.wells(name)) {
+                const auto otime_it = open_times.find(w);
+                if (otime_it == open_times.end()) {
+                    continue;
+                }
+                const double target_time = otime_it->second + wce.on_time;
+                if (target_time > current_time &&
+                    target_time < current_time + next_dt &&
+                    !opens_this_step(w))
```

Comment:
it's should be fairly cheap. currently it calls into the schedule, extracts the well events at the report step and checks if there is an opening event, https://github.com/OPM/opm-simulators/pull/5792/commits/86fb2f3d6116b1df4fc669a99633af67bc88c12f#diff-b17d026083f7ee29a4d9aa8ad05a54bda193055d6f313dd7bbc16f1d7ea83bf0R361

i cannot parse it up front because that would mean resolving the wlist up front which would be more costly i believe, and i didn't want to add an explicit circular dependency back to the schedule class.